custom_components/localtuya/__init__.py

```diff
@@ -246,29 +169,86 @@ async def _async_reconnect(now):
+        await discovery.start()
+        hass.data[DOMAIN][DATA_DISCOVERY] = discovery
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _shutdown)
+    except Exception:  # pylint: disable=broad-except
+        _LOGGER.exception("failed to set up discovery")
 
     return True
 
 
-async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
-    """Set up LocalTuya integration from a config entry."""
-    unsub_listener = entry.add_update_listener(update_listener)
+async def async_migrate_entry(hass, config_entry: ConfigEntry):
+    """Migrate old entries merging all of them in one."""
+    new_version = 2
+    if config_entry.version == 1:
+        _LOGGER.debug("Migrating config entry from version %s", config_entry.version)
+        stored_entries = hass.config_entries.async_entries(DOMAIN)
 
-    device = TuyaDevice(hass, entry.data)
+        if config_entry.entry_id == stored_entries[0].entry_id:
+            _LOGGER.debug(
+                "Migrating the first config entry (%s)", config_entry.entry_id
+            )
+            new_data = {}
+            new_data[CONF_REGION] = "eu"
+            new_data[CONF_CLIENT_ID] = "xxx"
+            new_data[CONF_CLIENT_SECRET] = "xxx"
+            new_data[CONF_USER_ID] = "xxx"
+            new_data[CONF_USERNAME] = DOMAIN
+            new_data[CONF_DEVICES] = {
+                config_entry.data[CONF_DEVICE_ID]: config_entry.data.copy()
+            }
+            new_data[ATTR_UPDATED_AT] = str(int(time.time() * 1000))
+            config_entry.version = new_version
+            hass.config_entries.async_update_entry(
+                config_entry, title=DOMAIN, data=new_data
+            )
+        else:
+            _LOGGER.debug(
+                "Merging the config entry %s into the main one", config_entry.entry_id
+            )
+            new_data = stored_entries[0].data.copy()
+            new_data[CONF_DEVICES].update(
+                {config_entry.data[CONF_DEVICE_ID]: config_entry.data.copy()}
+            )
+            new_data[ATTR_UPDATED_AT] = str(int(time.time() * 1000))
+            hass.config_entries.async_update_entry(stored_entries[0], data=new_data)
+            await hass.config_entries.async_remove(config_entry.entry_id)
+
+    _LOGGER.info(
+        "Entry %s successfully migrated to version %s.",
+        config_entry.entry_id,
+        new_version,
+    )
+    return True
 
-    hass.data[DOMAIN][entry.entry_id] = {
-        UNSUB_LISTENER: unsub_listener,
-        TUYA_DEVICE: device,
-    }
 
-    async def setup_entities():
-        platforms = set(entity[CONF_PLATFORM] for entity in entry.data[CONF_ENTITIES])
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up LocalTuya integration from a config entry."""
+    unsub_listener = entry.add_update_listener(update_listener)
+    hass.data[DOMAIN][UNSUB_LISTENER] = unsub_listener
+    hass.data[DOMAIN][TUYA_DEVICES] = {}
+
+    region = entry.data[CONF_REGION]
```

Comment:
Probably as I haven't *done* the cloud configuration to set up all of this; I get errors after upgrade.

May be worth raising errors if not .is_configured from a UX point of view
```
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 335, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/config/custom_components/localtuya/__init__.py", line 234, in async_setup_entry
    region = entry.data[CONF_REGION]
KeyError: 'region'
```